apps/user-cdc/src/main/java/it/pagopa/selfcare/user/event/repository/UserInstitutionRepository.java

```diff
@@ -0,0 +1,105 @@
+
+    private final UserMapper userMapper;
+
+    public void updateUser(UserInstitution userInstitution) {
+        OnboardedProductState state = retrieveStatusForGivenInstitution(userInstitution.getProducts());
+        UserInfo.findByIdOptional(userInstitution.getUserId())
+                .onItem().transformToUni(opt -> {
+                    if (VALID_PRODUCT_STATE.contains(state)) {
+                        PartyRole role = retrieveRoleForGivenInstitution(userInstitution.getProducts());
+                        return updateOrCreateNewUserInfo(opt, userInstitution, role, state);
+                    } else {
+                        return deleteInstitutionOrAllUserInfo(opt, userInstitution);
+                    }
+                })
+                .subscribe().with(
+                        result -> log.info("UserInfo collection successfully updated"),
+                        failure -> log.error("Error during UserInfo collection updating, message:" + failure.getMessage()));
+
+    }
+
+    private Uni<Void> deleteInstitutionOrAllUserInfo(Optional<ReactivePanacheMongoEntityBase> opt, UserInstitution userInstitution) {
+        return opt.map(UserInfo.class::cast)
+                .map(userInfo -> {
+                    if (userInfo.getInstitutions().stream()
+                            .anyMatch(userInstitutionRole -> userInstitutionRole.getInstitutionId().equalsIgnoreCase(userInstitution.getInstitutionId()))) {
+
+                        List<UserInstitutionRole> roleList = new ArrayList<>(userInfo.getInstitutions());
+                        roleList.removeIf(userInstitutionRole -> userInstitutionRole.getInstitutionId().equalsIgnoreCase(userInstitution.getInstitutionId()));
+                        userInfo.setInstitutions(roleList);
+
+                        if (roleList.isEmpty()) {
+                            return UserInfo.deleteById(userInstitution.getUserId()).replaceWith(Uni.createFrom().voidItem());
+                        } else {
+                            return UserInfo.persistOrUpdate(userInfo);
+                        }
+                    }
+                    return Uni.createFrom().voidItem();
+                })
+                .orElse(Uni.createFrom().voidItem());
+    }
+
+    private Uni<Void> updateOrCreateNewUserInfo(Optional<ReactivePanacheMongoEntityBase> opt, UserInstitution userInstitution, PartyRole role, OnboardedProductState state) {
+        return opt.map(UserInfo.class::cast)
+                .map(userInfo -> replaceOrAddInstitution(userInfo, userInstitution, role, state))
+                .orElse(userMapper.toNewUserInfo(userInstitution, role, state))
+                .persistOrUpdate()
+                .replaceWith(Uni.createFrom().voidItem());
+    }
+
+    private UserInfo replaceOrAddInstitution(UserInfo userInfo, UserInstitution userInstitution, PartyRole role, OnboardedProductState state) {
+        userInfo.getInstitutions().stream()
```

Comment:
you could check if is not null

Reply:
Institutions cannot be null by requirement. Do I check anyway?